tools/rosbag/src/player.cpp

```diff
@@ -335,6 +358,30 @@ void Player::doPublish(MessageInstance const& m) {
         bool charsleftorpaused = true;
         while (charsleftorpaused && node_handle_.ok())
         {
+            ros::spinOnce();
+
+            if (pause_change_requested_)
+            {
+              paused_ = requested_pause_state_;
+
+              if (paused_)
+              {
+                paused_time_ = ros::WallTime::now();
+              }
+              else
+              {
+                ros::WallDuration shift = ros::WallTime::now() - paused_time_;
```

Comment:
I realize this logic is duplicated in the keyboard input processing below, and can put it in a method if preferred.